### PR TITLE
[AIRFLOW-2499] Dockerised CI pipeline

### PR DESCRIPTION
**Pull request by gerardo**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2499
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Currently, running unit tests is a difficult process. Airflow tests depend on many external services and other custom setup, which makes it hard for contributors to work on this codebase. CI builds have also been unreliable[0][1][2][3], and it is hard to reproduce the causes. Having contributors trying to emulate the build environment every time makes it easier to get to an "it works on my machine" sort of situation.

This PR implements a dockerised version of the current build pipeline. This setup has a few advantages:

* TravisCI tests are reproducible locally
* The same build setup can be used to create a local development environment (there's a request for it [4])

### Implementation details

* I'm using Docker Compose for the container orchestration and configuration.
* MySQL, PostgreSQL, OpenLDAP, krb5 and rabbitmq are now services running inside their own containers
* I created a separate repo, called incubator-airflow-ci[5] (TravisCI build here[6]), where a base image with all dependencies is built. In this case, I'm following the same pattern the CouchDB[7] project follows
* Hadoop, Hive and MiniCluster were moved to this base image
* The current TravisCI pipeline lives here[8]

### How to run locally

* `docker-compose -f scripts/ci/docker-compose.yml run airflow-testing bash`
* Then, inside the container,: `tox -e py35-backend_sqlite` or any other valid `TOX_ENV` value

### Some observations

The whole environment setup on TravisCI (the setup before the tests are run) takes around 4 minutes. It is similar to the current version, but maybe some things can be optimised in the future by reducing the docker image size. Right now, the docker image size is hard to reduce, given we pre-install Hadoop, Hive and MiniCluster. You can check what contributes to the image size here[9]

I also had to change a few tests and default connections that were pointing to localhost. Now docker-compose takes care of creating an actual DNS name for the service containers, so the tests now refer to them by that name. The DNS names are the same as the container name in the docker-compose.yml file.

[0] https://issues.apache.org/jira/browse/AIRFLOW-671
[1] https://issues.apache.org/jira/browse/AIRFLOW-968
[2] https://issues.apache.org/jira/browse/AIRFLOW-2157
[3] https://issues.apache.org/jira/browse/AIRFLOW-2272
[4] https://issues.apache.org/jira/browse/AIRFLOW-1042 
[5] https://github.com/gerardo/incubator-airflow-ci
[6] https://travis-ci.org/gerardo/incubator-airflow-ci
[7] https://travis-ci.org/apache/couchdb-ci
[8] https://travis-ci.org/gerardo/incubator-airflow
[9] https://microbadger.com/images/gerardo/airflow-testing

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

It modifies unit tests. There are no new ones but the modified ones need revision.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### More

- [x] Fix impersonation tests
- [x] Re-Enable `Check License` build step
- [x] Document new CI / Testing environment
- [ ] Fix CodeCov
- [x] Delete travisci cache
- [x] Move https://github.com/gerardo/incubator-airflow-ci under Apache org

### CODE REVIEW

**Comment by gerardo** [comment]
The commit history look odd. Will fix it.

File: scripts/ci/krb5/Dockerfile
Comment (dimberman) [inline scripts/ci/krb5/Dockerfile]: We can't have the testing suite depend on a private docker-repo. Try to get this image hosted somewhere official like docker-repos.

File: scripts/ci/airflow_travis.cfg
Comment (dimberman) [inline scripts/ci/airflow_travis.cfg:58]: Definitely a fan of the more logical names :).

File: airflow/operators/python_operator.py
Comment (dimberman) [inline airflow/operators/python_operator.py]: are these python_operator changes necessary or can they be a seperate PR?

File: scripts/ci/docker-compose.yml
Comment (dimberman) [inline scripts/ci/docker-compose.yml]: Same issue here.

**Review by dimberman — CHANGES_REQUESTED** [review]
Overall I really like this idea. It'll make local testing significantly easier/standardize reduce issues with tests passing locally/failing on travis.

Definitely need a rebase to fix the commit history, and figure out a better solution for where the base image will live.

File: .travis.yml
Comment (rodrigc) [inline .travis.yml]: Instead of removing docker-compose, I would recommend upgrading it to the latest
via packages:
https://docs.docker.com/install/linux/docker-ce/debian/#set-up-the-repository

File: scripts/ci/krb5/Dockerfile
Comment (gerardo) [reply scripts/ci/krb5/Dockerfile]: Yeah, definitely agree on that. The idea is for the Airflow project to also have an `incubator-airflow-ci` repo as well, that generates this image semi-regularly under the `apache/airflow-testing` docker registry namespace, just like CouchDB and a few other Apache projects do https://travis-ci.org/apache/couchdb-ci.

The repo that generates that image lives under https://github.com/gerardo/incubator-airflow-ci, so you could consider it as part of this PR.

**Comment by dimberman** [comment]
That sounds like a really good idea. Would also go a long way towards
reducing run-time for tests. We'd have to figure out if offline development
is an issue though.

**Comment by dimberman** [comment]
```[error] 0.31% tests.ImpersonationTest.test_default_impersonation: [31m4.6534s[0m```
 Also looks like all of the impersionation tests are broken by this so should look into that.

File: airflow/operators/python_operator.py
Comment (gerardo) [reply airflow/operators/python_operator.py]: These tests were failing because the default pickling protocol for python 2 and 3 are different, so, if the python version being used to run airflow is different than the one passed to `PythonVirtualenvOperator` to create a virtualenv, then `pickle.load` fails. 

These tests are not failing on master, but they kept failing under this docker setup. Not sure if it's actually a bug in my setup or is it an actual issue in this operator, that's why I didn't create another  PR, but I can submit these changes as a separate PR if you think it's worth it.

Comment (dimberman) [reply airflow/operators/python_operator.py]: Making your PRs smaller is always recommended. You can reference this PR in that one.

File: .travis.yml
Comment (gerardo) [reply .travis.yml]: Yeah, makes sense, I was also going by the TravisCI documentation: https://docs.travis-ci.com/user/docker/#Using-Docker-Compose. Not sure if there is an actual reason they prefer this method over adding the repo.

**Comment by gerardo** [comment]
@dimberman re: offline development, docker images are cached locally once you download them. If you know you'll be offline, you can download the image beforehand (running `docker-compose -f scripts/ci/docker-compose.yml run airflow-testing bash` once should be enough to trigger this). 

There's also a  `docker save` command to save this image to a tarfile, so that it can be imported later.

**Comment by dimberman** [comment]
Fair point. Yeah, overall I think this is a great idea. I've had multiple python projects switch to docker based testing and it's helped a lot. Would also drastically lower the barrier for new contributors (but please clean up your commits. You can definitely squash this down a lot).

**Comment by gerardo** [comment]
@dimberman re: impersonation tests, are you familiar with the preconditions for impersonation in general to work? These tests have been hard to debug, and I'm still not sure how to fix them. The only reason I can think of they might not be working is that I'm running the main container with the root user, instead of a regular user with superpowers.

File: tox.ini
Comment (milton0825) [inline tox.ini]: probably need to uncomment this.

Comment (milton0825) [reply tox.ini]: nvm. looks like its in your check list.

**Comment by milton0825** [comment]
I am looking forward to having this PR merged! The requirement for running the unit test should be simple and straightforward.

**Comment by gerardo** [comment]
Hello @dimberman @milton0825 @rodrigc 

So, I managed to address a few outstanding issues with this build. Currently the tests are failing because master is broken though (test_check_operators was broken a few commits ago it seems). In the meantime, I think this is ready. I would love if you guys can try it out.

I'm concerned about documentation. A few days ago, Fokko introduced a little bit of documentation for [setting up a development environment using docker](https://github.com/apache/incubator-airflow/commit/7db18183677d96e37f4671aa846815d3ecc7e58c). It only adds a section to use a docker container for dev mode, but it doesn't address the rest of the contributor's documentation, e.g:

* `./run_unit_tests.sh` is completely unusable
*  There are lots of integration tests in the test suite, which need a more comprehensive setup (MySQL, krb5 tests) than just a single docker container with python
* And of course, calling these "unit tests" is simply not accurate. These are integration tests as far as I can tell

For now, I decided to document [3 ways of setting up a dev environment](https://github.com/gerardo/incubator-airflow/blob/dd7a33ae4493b10332e76c85e000ea23274d89c8/CONTRIBUTING.md#development-and-testing). I think it might be better to just use this PR to address the CI environment, and create a separate one to document a development workflow properly.

What do you think?

**Comment by milton0825** [comment]
@gerardo I am seeing master pass tests on TravisCI. Can you rebase master and retry? The commiters won't merge it unless tests pass on CI tho.

**Comment by gerardo** [comment]
@milton0825 where did you see that? I just checked https://travis-ci.org/apache/incubator-airflow/branches and the last five builds of master are red.

**Comment by gerardo** [comment]
PR https://github.com/apache/incubator-airflow/pull/3485 is waiting to get merged and fix the current CI issue.

**Comment by milton0825** [comment]
nvm. I must be looking at the wrong repo.

**Comment by Noremac201** [comment]
Just a sidenote,
Running on an operating system without /bin in $PATH gives a ``[FATAL tini (8)] exec bash failed: No such file or directory`` error. For example, (unless I've misconfigured up my OS) I don't believe Arch linux has /bin in $PATH by default.

I fixed it by adding 
``- PATH=${PATH}:/bin``
to the docker-compose.yaml, under the similar command for ``/usr/bin/hive``

**Comment by gerardo** [comment]
@Noremac201 thanks for checking this out! That line is meant to add hive to the path, but from inside the container, where /bin should've been already added to the path. The bug there is that `$PATH` is referring to the PATH of the host, and it should refer to the container one. 

I'll fix it and will let you know so you can try it again.

**Comment by gerardo** [comment]
An update on this PR:

- Hive is added to `$PATH` on the [base image](https://github.com/gerardo/incubator-airflow-ci) now instead of the docker-compose file. Thanks @Noremac201 for reporting this out.
- A recently merged PR introduced Cassandra as a dependency. It was really simple to add a new Casandra service on docker-compose file and have the default `CassandraHook` point to the `cassandra` host 😄 
- I'm also building [my branch on TravisCI](https://travis-ci.com/gerardo/incubator-airflow/), which is passing. Through this, I found a problem with the pip cache under the TravisCI Apache org. The current pip cache needs to be deleted once as part of merging this PR. The `mysqlclient` library there was built against the travisci host (ubuntu trusty), but the docker-compose version is running ubuntu xenial (a more recent version). That causes the [current build failures in this branch](https://travis-ci.org/apache/incubator-airflow/builds/394426183).

If someone could delete the cache, that would be great (I guess I could just `rm -rf ~/.cache/pip` from the travisci script, push that change, wait for it to run, then remove it)

@rodrigc @dimberman @milton0825 @Fokko @r39132 PTAL

**Comment by gerardo** [comment]
I went ahead and deleted the cache myself. The tests are green now. I wonder if someone can have a look?

**Comment by milton0825** [comment]
LGTM. I think we should push this out ASAP as it really improves the development productivity. Also, it makes other people more willing to contribute as the testing setup is really painful in the past.

**Comment by milton0825** [comment]
@Fokko @kaxil

**Comment by bolkedebruin** [comment]
Nice looks great! @Fokko @jgao54 @criccomini

**Review by Fokko — APPROVED** [review]
I like this a lot, this also enables you to run, for example, the hive tests locally 👍

File: scripts/ci/docker-compose.yml
Comment (dimberman) [inline scripts/ci/docker-compose.yml]: Please change to non-personal repo

**Review by dimberman — CHANGES_REQUESTED** [review]

**Comment by dimberman** [comment]
@gerardo  Please let me know if there is anything I can do to help. I really want this feature :)

**Comment by dimberman** [comment]
@bolkedebruin @fokko could we keep a base image somewhere to pull from? If we had an image with all expected libraries pre-installed it would cut image build times significantly.

File: scripts/ci/docker-compose.yml
Comment (gerardo) [reply scripts/ci/docker-compose.yml]: @dimberman I'll change it to `apache/airflow-testing`, but we need an image living there first. The repo with the current `Dockerfile` lives in https://github.com/gerardo/incubator-airflow-ci and here's the current TravisCI build: https://travis-ci.org/gerardo/incubator-airflow-ci. 

This should live under the `apache` org

Comment (gerardo) [reply scripts/ci/docker-compose.yml]: BTW, this is the same process used by the CouchDB project: https://travis-ci.org/apache/couchdb-ci and a few others under the `apache` org.

**Comment by gerardo** [comment]
Hi @Fokko @dimberman @bolkedebruin. I'm following up on this PR. What else do you think needs to be fixed before merging this PR?

I just rebased from master. It's becoming a little bit painful to merge changes and fix changes from other merged PRs that add new dependencies using the current CI setup. For example, I had to reimplement https://github.com/apache/incubator-airflow/pull/3611 to make it work with Docker Compose. Even though it was rather simple, these sort of changes can add up over time.

**Comment by Fokko** [comment]
Hi @gerardo, I really appreciate the effort that you're putting into this. I think the last step is creating the `incubator-airflow-ci` repository to get the Dockerbuild in there. I've requested a new respository to build the docker image: https://issues.apache.org/jira/browse/INFRA-16905

File: .travis.yml
Comment (Fokko) [inline .travis.yml]: Nice one, this one is now installed by pip: https://github.com/apache/incubator-airflow/pull/3691

File: .travis.yml
Comment (Fokko) [inline .travis.yml]: Maybe get the version of of there? Like: https://docs.travis-ci.com/user/docker/#using-docker-compose

File: scripts/ci/docker-compose.yml
Comment (Fokko) [reply scripts/ci/docker-compose.yml]: Requested under https://issues.apache.org/jira/browse/INFRA-16905

Comment (Fokko) [reply scripts/ci/docker-compose.yml]: Hi Gerardo, it took a while. But the docker image is being build here: https://hub.docker.com/r/airflowci/incubator-airflow-ci/builds/

Based on this repository: https://github.com/apache/incubator-airflow-ci

Comment (Fokko) [reply scripts/ci/docker-compose.yml]: @gerardo Could you update the PR?

**Comment by gerardo** [comment]
@Fokko Thanks for the reviews. I applied the requested changes and rebased against master. Also thanks for setting up [apache/incubator-airflow-ci](https://github.com/apache/incubator-airflow-ci).

File: .travis.yml
Comment (Fokko) [inline .travis.yml:48]: One small nit, if we use the `docker` service on Travis, we get docker-compose for free: https://docs.travis-ci.com/user/docker/

**Review by Fokko — CHANGES_REQUESTED** [review]
@gerardo I still have one nit, not blocking btw.

@kaxil @feng-tao Can you also take a look? I'll check if the code coverage is still working. Right now, it looks like the report is not being uploaded: https://codecov.io/gh/apache/incubator-airflow/pull/3393

File: .travis.yml
Comment (gerardo) [reply .travis.yml:48]: @Fokko Ah yeah, docker-compose. There's a reason for this. 

In earlier versions of this PR, the build logs were massive, because, by default, docker-compose shows ascii progress bars when downloading docker images. These progress bars don't render properly in TravisCI (it becomes like 100 lines per progress bar). The only way to disable this is to [set the docker-compose log-level to `ERROR`](https://github.com/apache/incubator-airflow/pull/3393/files#diff-354f30a63fb0907d4ad57269548329e3R55).

This change was only added in docker-compose [1.20.0](https://github.com/docker/compose/blob/master/CHANGELOG.md#all-formats-1). The TravisCI version is very old (1.4 if I remember correctly).

File: run_unit_tests.sh
Comment (kaxil) [inline run_unit_tests.sh:24]: Nothing major but can I know why we selected this instead of `~/airflow`?

**Comment by kaxil** [comment]
This is great work... 🎉  Thanks @gerardo for picking this up.

File: run_unit_tests.sh
Comment (gerardo) [reply run_unit_tests.sh:24]: @kaxil after I moved everything to docker, I realized it makes more sense for `AIRFLOW_HOME` to be the actual `HOME` directory that belongs to the `airflow` user. 

`~/airflow` makes sense if there are more files in the same home directoy (for example, when running locally). But that's not the case anymore in a dockerised environment.

File: .travis.yml
Comment (Fokko) [reply .travis.yml:48]: 👍 Makes sense to me

**Review by Fokko — APPROVED** [review]

**Comment by Fokko** [comment]
LGTM, Thanks @gerardo

**Comment by dimberman** [comment]
🙌🙌🙌🙌🙌🙌🙌🙌

File: .travis.yml
Comment (kaxil) [inline .travis.yml:32]: @gerardo I am currently updating the Airflow CI to point to 3.6 instead of python 3.5 

But the way Python Version is passed using this environment variable, can you help me figure out how this can be achieved?

Comment (kaxil) [reply .travis.yml:32]: Also, we need to keep in mind that in future we would need to test both Python3.6 and Python3.7

Comment (kaxil) [reply .travis.yml:32]: https://github.com/apache/incubator-airflow/pull/3815

Comment (gerardo) [reply .travis.yml:32]: @kaxil give we're using tox for this, it should be a matter of just adding/updating the runtimes inside the tox file [here](https://github.com/apache/incubator-airflow/blob/master/tox.ini#L20) and [here](https://github.com/apache/incubator-airflow/blob/master/tox.ini#L40) and then pass the desired python version through the Travis script as we do now.

There's an extra check inside the [run-ci script](https://github.com/apache/incubator-airflow/blob/master/scripts/ci/run-ci.sh#L29) to use the correct pip binary (pip vs pip3), but this shouldn't be affected, as that only cares about the mayor version.

**Pull request merged**
